gowin/gfx.h

```diff
@@ -0,0 +1,4935 @@
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#ifndef GFX_H
+#define GFX_H
+
+#include "nextpnr.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+// LUTs
+const float lut_w = 0.6732 - 0.6386;
+const float lut_h = 0.9392 - 0.9074;
+const float lut_x = 0.6386;
+const float lut_y[] = {1. - 0.9392, 1. - 0.8870, 1. - 0.7864, 1. - 0.7321,
+                       1. - 0.6399, 1. - 0.5847, 1. - 0.5068, 1. - 0.4503};
+const float dff_w = 0.0580;
+const float dff_x = 0.6821;
+const float grp_lut_w = 0.1399;
+const float grp_lut_h = 0.0954;
+const float grp_lut_x = 0.6284;
+const float grp_lut_y[] = {1. - 0.9440, 1. - 0.7915, 1. - 0.6442, 1. - 0.5101};
+
+// mux
+const float mux_w = 0.8134 - 0.7899;
+const float mux_f = 0.9450 - 0.9358;
+const float mux_h = grp_lut_h;
+
+const float mux2lut5_x = 0.7900;
+const float mux2lut5_y[] = {grp_lut_y[0], grp_lut_y[1], grp_lut_y[2], grp_lut_y[3]};
+const float mux2lut6_x = 0.8378;
+const float mux2lut6_y[] = {1. - 0.9261, 1. - 0.6205};
+const float mux2lut7_x = 0.8859;
+const float mux2lut7_y = 1. - 0.7870;
+const float mux2lut8_x = 0.9337;
+const float mux2lut8_y = 1. - 0.8098;
+
+// pip
+enum CruSide
+{
+    Top,
+    Bottom,
+    Left,
+    Right,
+    Center
+};
+const float cru_x = 0.2568;
+const float cru_y = 1. - 0.9783;
+const float cru_w = 0.6010 - cru_x;
+const float cru_h = 1. - cru_y - 0.3742;
+
+const float lut_A_off = 1. - 0.9107 - lut_y[0];
+const float lut_D_off = lut_h - lut_A_off;
+const float lut_B_off = lut_A_off - (lut_h - lut_D_off) / 3.;
+const float lut_C_off = lut_D_off + (lut_h - lut_D_off) / 3.;
+
+const float right_wire_dist = (grp_lut_y[1] - grp_lut_y[0] - grp_lut_h) / 11.;
+const float left_wire_dist = cru_h / 100.;
+const float top_wire_dist = cru_w / 100.;
+const float clk_ce_set_vdist = (lut_y[1] - lut_y[0] - lut_h) / 4.;
+
+const float sn_dist = cru_x / 125.;
+const float ew_dist = (1. - cru_y - cru_h) / 130.;
+const float wrap_len = 0.02f;
+const float spine_pip_off = 0.11f;
+
+const float io_x = cru_x + cru_w + 0.1;
+const float io_w = (1. - io_x) / 3.;
+const float io_gap = 0.03;
+const float io_h = (cru_h - 4. * io_gap) / 2.;
+const float io_y = cru_y + io_gap;
+
+const float ios_scl = 0.5;
+const float ios_h = ios_scl * io_h;
+const float ios_w = ios_scl * io_w;
+const float ios_gap_y = io_gap;
+const float ios_gap_x = io_gap * 1.4;
+const float ios_x = io_x;
+const float ios_y = ios_scl * io_y;
+
+const dict<IdString, float> portPoint = {
+        {id_O, 3. * io_h / 4.},
+        {id_I, 2. * io_h / 4.},
+        {id_OEN, 1. * io_h / 4.},
+};
+
+const dict<IdString, std::vector<std::tuple<float, float, float, float>>> portSign = {
```

Comment:
out of curiosity, were these autogenerated by something? would it be possible to include the generator script for reference, too?

Reply:
Unfortunately, there is no generator, I write this way by hand:(